Added initial support for gcc's gomp implementation.

oodler577 opened the pull request:
cflags   -> -fopenmp
  ldflags  -> -fopenmp

Makefile.PL also uses Devel::CheckLib to verify we're
in a sane gcc environment when ccname == 'gcc'.

mohawk2 commented:
This is a great start! A few points:
- Devel::CheckLib will need to be a `CONFIGURE_REQUIRES`
- the CI will need to pass, and probably the Ubuntu package that needs adding is `libomp-dev`
- it's not acceptable for the module to install (or indeed succeed in the configure phase) if the environment can't do OpenMP, so the `if gcc` bit (which obviously can be expanded for other possibilities) needs an `else` that does `os_unsupported`
- the `check_lib` bit will need a `function` as previously discussed to ensure actual OpenMP per the standard (that could go in the `else` as a Hail Mary) - this implies starting with a `%check_lib_args` of defaults which get added to in the `if` branches, then `check_lib` gets run once with it (achieving DRY)
- the test shows a code-driven approach, and a more sustainable one is to separate the environment-info-gathering phase, from the info-interpretation phase; with that, a test could run anywhere since it would just have a hardcoded list of infos, and do `is_deeply` on the interpretations

oodler577 commented:
Note on one of your comments above, @mohawk2 -

> the test shows a code-driven approach, and a more sustainable one is to separate the environment-info-gathering phase, from the info-interpretation phase; with that, a test could run anywhere since it would just have a hardcoded list of infos, and do is_deeply on the interpretations

If you are suggesting maintaining a set of profiles, I'd prefer to way until we might actually need it. The current approach simply updates a localized `%Config::Config` by inserting `compiler_name` as desired. I am open to expanding this in either case. The test is meant to be decoupled from the environment. If you can show me how it's not, I'd be happy make sure that it's fully independent as such. Since I do agree with your main point, which I read as, "the tests should run everywhere regardless of the environment."

mohawk2 reviewed (changes requested):
I don't think all the points mentioned in my comment on the main PR have been addressed yet? Further thoughts:
- An example to copy for installing Ubuntu packages: https://github.com/PDLPorters/perl5-PGPLOT/blob/master/.github/workflows/ci.yml#L45
- It makes a reviewer's life easier if you avoid unnecessary whitespace diffs

oodler577 commented:
still working down the list:)

oodler577 commented:
@mohawk2 I expect the changes to pass. I believe I have addressed all issues you presented. Please let me know if there is anything else. I'd like to add you under 'AUTHORS', please tell me how you'd like that listed as one final change.

oodler577 commented:
@mohawk2 - no rush, but I'd like to get these changes finalized today. I am eager to begin using this modules for other things. I also have quite a few "Alien" modules I think I might be able to knock out for (HPC related mostly). Anyway, just wanted to finish so I can recover any bandwidth and even utlize this for the talk I am preparing for the end of the month. Thanks for your continued guidence!

mohawk2 commented on lib/Alien/OpenMP.pm:
```diff
@@ -15,15 +43,80 @@ Alien::OpenMP - Encapsulate system info for OpenMP
+initial support is targeting it. However, like C<perl>, many other compilers
+may be used.
+
+Adding support for a new compiler should be straightforward; while C<pull
```
I'd have thought this should be an `I<>` or nothing at all, it's not code? :-)

mohawk2 commented on lib/Alien/OpenMP.pm:
```diff
@@ -15,15 +43,80 @@ Alien::OpenMP - Encapsulate system info for OpenMP
+
+If used for an unsupported compiler, C<ExtUtils::MakeMaker::os_unsupported> is
+invoked, which results an exception propagating from this method being raised
+with the value of C<qq{OS unsupported\n}> (note the new line).
```
This is not important detail. The user notionally has to read it, or skim it to learn if they can ignore it. Also, you had to write it. What does it add that "won't install if not valid" didn't?

mohawk2 commented on t/01-basic.t:
```diff
@@ -0,0 +1,55 @@
+# add additional tests as compiler support is added to
+# this module.
+
+GCC:
```
My suggestion was of a separate phases of info-gather and info-interpret, which would then be testable with `is_deeply`. This is still code.

mohawk2 commented on xt/pod.t:
```diff
@@ -1,4 +1,6 @@
-use strict; use warnings;
+use strict;
```
What is the benefit of changing this? A reviewer has to read through these diffs before being able to conclude they are semantically null.

mohawk2 commented on Makefile.PL:
```diff
@@ -1,62 +1,99 @@
+        ccflags => q{-fopenmp},
+        ldflags => q{-fopenmp},
+        function => <DATA>,    # main() body defined in __DATA__
+    );
```
This is still not a `%check_lib_args` defined above the `if` block, modified within the `if` block, and one `check_lib_or_exit` call in the whole program. What would the code look like if you supported a second environment?

oodler577 replied:
Oh, you're saying to set it up to be easily extended in the future. Will do, now that I understand what you mean :)

mohawk2 commented on .github/workflows/ci.yml:
```diff
@@ -46,7 +50,7 @@ jobs:
       fail-fast: false
       matrix:
-        os: [macos-latest, windows-latest]
+        os: [windows-latest]
```
Is there a way to also support MacOS? See https://stackoverflow.com/questions/43555410/enable-openmp-support-in-clang-in-mac-os-x-sierra-mojave

oodler577 replied:
yes, will do

mohawk2 reviewed (commented):
> Note on one of your comments above, @mohawk2 -
> > the test shows a code-driven approach, and a more sustainable one is to separate the environment-info-gathering phase, from the info-interpretation phase; with that, a test could run anywhere since it would just have a hardcoded list of infos, and do is_deeply on the interpretations
> 
> If you are suggesting maintaining a set of profiles, I'd prefer to way until we might actually need it. The current approach simply updates a localized `%Config::Config` by inserting `compiler_name` as desired.

I can see it does that. I am advocating you change it.

> I am open to expanding this in either case. The test is meant to be decoupled from the environment. If you can show me how it's not, I'd be happy make sure that it's fully independent as such. Since I do agree with your main point, which I read as, "the tests should run everywhere regardless of the environment."

The situation so far is that indeed is! All we're discussing is how. I have yet to see a situation in computing where code-driven is better than data-driven; nearly all situations I've seen have had data-driven approaches be much easier, and much better.

mohawk2 commented:
There is still no separate info-gather and info-evaluating phases which would enable the "basic" test to be data-driven. I will merge and release this anyway.
